### PR TITLE
Various HttpListener managed implementation improvements

### PR DESCRIPTION
**Pull request by hughbe**
Contributes to #18128

### CODE REVIEW

File: src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs [hunk 1]
```diff
@@ -313,7 +313,6 @@ public void Dispose()
         }
 
         [Fact]
-        [ActiveIssue(18128, platforms: TestPlatforms.AnyUnix)] // NotSupportedException thrown instead of InvalidOperationException
```
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: How did overriding FlushAsync fix this? The override is fine to add, but without it, it should have just used the base implementation, which would have queued a call to Flush.

File: src/System.Net.HttpListener/src/System/Net/Managed/ChunkedInputStream.cs [hunk 2]
```diff
@@ -76,7 +76,7 @@ public override int Read(byte[] buffer, int offset, int count)
             return EndRead(ares);
         }
 
-        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback cback, object state)
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int size, AsyncCallback cback, object state)
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Managed/ChunkedInputStream.cs]: Why did you change this from count to size?  The base implementation uses "count"... I don't see a good reason to move away from that.

File: src/System.Net.HttpListener/src/System/Net/Managed/HttpRequestStream.Managed.cs [hunk 3]
```diff
@@ -130,13 +130,25 @@ private int FillFromBuffer(byte[] buffer, int offset, int count)
             return size;
         }
 
-        public override int Read(byte[] buffer, int offset, int count)
+        public override int Read(byte[] buffer, int offset, int size)
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Managed/HttpRequestStream.Managed.cs]: Same question

File: src/System.Net.HttpListener/src/System/Net/Managed/HttpRequestStream.Managed.cs [hunk 4]
```diff
@@ -146,24 +158,36 @@ public override int Read(byte[] buffer, int offset, int count)
...
 
-            nread = _stream.Read(buffer, offset, count);
+            nread = _stream.Read(buffer, offset, size);
             if (nread > 0 && _remainingBody > 0)
                 _remainingBody -= nread;
             return nread;
         }
 
-        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback cback, object state)
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int size, AsyncCallback cback, object state)
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Managed/HttpRequestStream.Managed.cs]: Same question... will stop commenting on this.

File: src/System.Net.HttpListener/src/System/Net/Managed/ChunkedInputStream.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Managed/ChunkedInputStream.cs]: Oh, I see, the existing Windows implementation uses "size" and that gets encoded into the argument of the exception that's thrown.  Ugh.  Ok.

File: src/System.Net.HttpListener/tests/HttpResponseStreamTests.cs [hunk 5]
```diff
@@ -478,8 +478,7 @@ public void Dispose()
 
                 // Disconnect the Socket from the HttpListener.
                 client.Close();
-                GC.Collect();
-                GC.WaitForPendingFinalizers();
+                Thread.Sleep(2000);
```
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpResponseStreamTests.cs]: This isn't reliable.  We either need to find a way to coordinate between the client and server to ensure the proper behavior, or delete the test.

File: src/System.Net.HttpListener/tests/HttpResponseStreamTests.cs [hunk 6]
```diff
@@ -520,8 +519,7 @@ public void Dispose()
...
```
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpResponseStreamTests.cs]: Ditto

**Comment by stephentoub** [comment]
Generally looks good.  Thanks for working on this, @hughbe.

File: src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: It fixes another assertion failure in the test method, There were multiple issues here: https://github.com/hughbe/corefx/blob/b3b47377337e1cb5532443fc63326d82b91a518e/src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs#L329

Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: So the comment next to the attribute was wrong?

Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: It was correct because the other assertions in the test were failing, but also partially wrong as it didn't mention anything about flushasync, I think.

Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: I'm just confused because the only change to the src being made in this commit is the change to FlushAsync, which shouldn't have any affect on a NotSupportedException vs InvalidOperationException.  But maybe the comment was stale and that NSE vs IOE issue was previously fixed.  Regardless, if it's passing now, great.

File: src/System.Net.HttpListener/tests/HttpResponseStreamTests.cs (hunk 5 above)
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpResponseStreamTests.cs]: (If you wanted to remove this commit, I think the rest of the change is good to go in.)

File: src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: I think that issue was fixed in https://github.com/dotnet/corefx/pull/18450/commits/5858c5b4b3d438d17aec8090906a62a9e87269e7
I just undisabled the test too early!

File: src/System.Net.HttpListener/tests/HttpResponseStreamTests.cs (hunk 5 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpResponseStreamTests.cs]: Yup, reverted this commit.
I'm willing to delete these if there's *really* no way to check this. But it does test an important piece of functionality of HttpListener - the `IgnoreWriteExceptions` feature
